File: app/src/main/java/net/ktnx/mobileledger/backup/RawConfigWriter.java
/*
 * Copyright © 2021 Damyan Ivanov.
 * This file is part of MoLe.
 * MoLe is free software: you can distribute it and/or modify it
 * under the term of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your opinion), any later version.
 *
 * MoLe is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License terms for details.
 *
 * You should have received a copy of the GNU General Public License
 * along with MoLe. If not, see <https://www.gnu.org/licenses/>.
 */

package net.ktnx.mobileledger.backup;

import android.util.JsonWriter;

import net.ktnx.mobileledger.backup.ConfigIO.Keys;
import net.ktnx.mobileledger.db.Currency;
import net.ktnx.mobileledger.db.DB;
import net.ktnx.mobileledger.db.Profile;
import net.ktnx.mobileledger.db.TemplateAccount;
import net.ktnx.mobileledger.db.TemplateWithAccounts;
import net.ktnx.mobileledger.json.API;
import net.ktnx.mobileledger.model.Data;

import java.io.BufferedWriter;
import java.io.IOException;
import java.io.OutputStream;
import java.io.OutputStreamWriter;
import java.util.List;

public class RawConfigWriter {
    private final JsonWriter w;
    public RawConfigWriter(OutputStream outputStream) {
        w = new JsonWriter(new BufferedWriter(new OutputStreamWriter(outputStream)));
        w.setIndent("  ");
    }
    public void writeConfig() throws IOException {
        w.beginObject();
        writeCommodities();
        writeProfiles();
        writeCurrentProfile();
        writeConfigTemplates();
        w.endObject();
        w.flush();
    }
    private void writeKey(String key, String value) throws IOException {
        if (value != null)
            w.name(key)
             .value(value);
    }
    private void writeKey(String key, Integer value) throws IOException {
        if (value != null)
            w.name(key)
             .value(value);
    }
    private void writeKey(String key, Long value) throws IOException {
        if (value != null)
            w.name(key)
             .value(value);
    }
    private void writeKey(String key, Float value) throws IOException {
        if (value != null)
            w.name(key)
             .value(value);
    }
    private void writeKey(String key, Boolean value) throws IOException {
        if (value != null)
            w.name(key)
             .value(value);
    }
    private void writeConfigTemplates() throws IOException {
        List<TemplateWithAccounts> templates = DB.get()
                                                 .getTemplateDAO()
                                                 .getAllTemplatesWithAccountsSync();

        if (templates.isEmpty())
            return;

        w.name("templates")
         .beginArray();
        for (TemplateWithAccounts t : templates) {
            w.beginObject();

            w.name(Keys.UUID)
             .value(t.header.getUuid());
            w.name(Keys.NAME)
             .value(t.header.getName());
            w.name(Keys.REGEX)
             .value(t.header.getRegularExpression());
            writeKey(Keys.TEST_TEXT, t.header.getTestText());
            writeKey(ConfigIO.Keys.DATE_YEAR, t.header.getDateYear());
            writeKey(Keys.DATE_YEAR_GROUP, t.header.getDateYearMatchGroup());
            writeKey(Keys.DATE_MONTH, t.header.getDateMonth());
            writeKey(Keys.DATE_MONTH_GROUP, t.header.getDateMonthMatchGroup());
            writeKey(Keys.DATE_DAY, t.header.getDateDay());
            writeKey(Keys.DATE_DAY_GROUP, t.header.getDateDayMatchGroup());
            writeKey(Keys.TRANSACTION, t.header.getTransactionDescription());
            writeKey(Keys.TRANSACTION_GROUP, t.header.getTransactionDescriptionMatchGroup());
            writeKey(Keys.COMMENT, t.header.getTransactionComment());
            writeKey(Keys.COMMENT_GROUP, t.header.getTransactionCommentMatchGroup());
            w.name(Keys.IS_FALLBACK)
             .value(t.header.isFallback());
            if (t.accounts.size() > 0) {
                w.name(Keys.ACCOUNTS)
                 .beginArray();
                for (TemplateAccount a : t.accounts) {
                    w.beginObject();

                    writeKey(Keys.NAME, a.getAccountName());
                    writeKey(Keys.NAME_GROUP, a.getAccountNameMatchGroup());
                    writeKey(Keys.COMMENT, a.getAccountComment());
                    writeKey(Keys.COMMENT_GROUP, a.getAccountCommentMatchGroup());
                    writeKey(Keys.AMOUNT, a.getAmount());
                    writeKey(Keys.AMOUNT_GROUP, a.getAmountMatchGroup());
                    writeKey(Keys.NEGATE_AMOUNT, a.getNegateAmount());
                    writeKey(Keys.CURRENCY, a.getCurrency());
                    writeKey(Keys.CURRENCY_GROUP, a.getCurrencyMatchGroup());

                    w.endObject();
                }
                w.endArray();
            }

            w.endObject();
        }
        w.endArray();
    }
    private void writeCommodities() throws IOException {
        List<Currency> list = DB.get()
                                .getCurrencyDAO()
                                .getAllSync();
        if (list.isEmpty())
            return;
        w.name(Keys.COMMODITIES)
         .beginArray();
        for (Currency c : list) {
            w.beginObject();
            writeKey(Keys.NAME, c.getName());
            writeKey(Keys.POSITION, c.getPosition());
            writeKey(Keys.HAS_GAP, c.getHasGap());
            w.endObject();
        }
        w.endArray();
    }
    private void writeProfiles() throws IOException {
        List<Profile> profiles = DB.get()
                                   .getProfileDAO()
                                   .getAllOrderedSync();

        if (profiles.isEmpty())
            return;

        w.name(Keys.PROFILES)
         .beginArray();
        for (Profile p : profiles) {
            w.beginObject();

            w.name(Keys.NAME)
             .value(p.getName());
            w.name(Keys.UUID)
             .value(p.getUuid());
            w.name(Keys.URL)
             .value(p.getUrl());
            w.name(Keys.USE_AUTH)
             .value(p.isAuthEnabled());
            if (p.isAuthEnabled()) {
                w.name(Keys.AUTH_USER)
                 .value(p.getAuthUser());
                w.name(Keys.AUTH_PASS)
                 .value(p.getAuthPassword());
            }
            if (p.getApiVersion() != API.auto.toInt())
                w.name(Keys.API_VER)
                 .value(p.getApiVersion());
            w.name(Keys.CAN_POST)
             .value(p.canPost());
            if (p.canPost()) {
                String defaultCommodity = p.getDefaultCommodityOrEmpty();
                if (!defaultCommodity.isEmpty())
                    w.name(Keys.DEFAULT_COMMODITY)
                     .value(defaultCommodity);
                w.name(Keys.SHOW_COMMODITY)
                 .value(p.getShowCommodityByDefault());
                w.name(Keys.SHOW_COMMENTS)
                 .value(p.getShowCommentsByDefault());
                w.name(Keys.FUTURE_DATES)
                 .value(p.getFutureDates());
                w.name(Keys.PREF_ACCOUNT)
                 .value(p.getPreferredAccountsFilter());
            }
            w.name(Keys.COLOUR)
             .value(p.getTheme());

            w.endObject();
        }
        w.endArray();
    }
    private void writeCurrentProfile() throws IOException {
        Profile currentProfile = Data.getProfile();
        if (currentProfile == null)
            return;

        w.name(Keys.CURRENT_PROFILE)
         .value(currentProfile.getUuid());
    }
}

File: app/src/main/kotlin/net/ktnx/mobileledger/backup/RawConfigWriter.kt
/*
 * Copyright © 2021 Damyan Ivanov.
 * This file is part of MoLe.
 * MoLe is free software: you can distribute it and/or modify it
 * under the term of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your opinion), any later version.
 *
 * MoLe is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License terms for details.
 *
 * You should have received a copy of the GNU General Public License
 * along with MoLe. If not, see <https://www.gnu.org/licenses/>.
 */

package net.ktnx.mobileledger.backup

import java.io.IOException
import java.io.OutputStream
import kotlin.coroutines.coroutineContext
import kotlinx.coroutines.ensureActive
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.buildJsonArray
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put
import net.ktnx.mobileledger.core.database.entity.TemplateAccount
import net.ktnx.mobileledger.core.database.entity.TemplateWithAccounts
import net.ktnx.mobileledger.core.domain.model.Currency
import net.ktnx.mobileledger.core.domain.model.Profile
import net.ktnx.mobileledger.core.domain.repository.CurrencyRepository
import net.ktnx.mobileledger.core.domain.repository.ProfileRepository
import net.ktnx.mobileledger.domain.repository.TemplateRepository
import net.ktnx.mobileledger.json.API

/**
 * Writes MoLe configuration to a backup file using kotlinx-serialization.
 *
 * Uses explicit JSON construction to match the original backup format with
 * conditional field writing.
 */
class RawConfigWriter(
    private val outputStream: OutputStream,
    private val profileRepository: ProfileRepository,
    private val templateRepository: TemplateRepository,
    private val currencyRepository: CurrencyRepository
) {
    private val prettyJson = Json { prettyPrint = true }

    @Throws(IOException::class)
    suspend fun writeConfig() {
        coroutineContext.ensureActive()

        val jsonObject = buildJsonObject {
            // Write commodities
            val commodities = getCommodities()
            if (commodities.isNotEmpty()) {
                put(
                    BackupKeys.COMMODITIES,
                    buildJsonArray {
                        commodities.forEach { add(buildCommodityJson(it)) }
                    }
                )
            }

            coroutineContext.ensureActive()

            // Write profiles
            val profiles = getProfiles()
            if (profiles.isNotEmpty()) {
                put(
                    BackupKeys.PROFILES,
                    buildJsonArray {
                        profiles.forEach { add(buildProfileJson(it)) }
                    }
                )
            }

            coroutineContext.ensureActive()

            // Write current profile
            val currentProfile = profileRepository.currentProfile.value
            if (currentProfile != null) {
                put(BackupKeys.CURRENT_PROFILE, currentProfile.uuid)
            }

            coroutineContext.ensureActive()

            // Write templates
            val templates = getTemplates()
            if (templates.isNotEmpty()) {
                put(
                    BackupKeys.TEMPLATES,
                    buildJsonArray {
                        templates.forEach { add(buildTemplateJson(it)) }
                    }
                )
            }
        }

        val jsonString = prettyJson.encodeToString(JsonObject.serializer(), jsonObject)
        val writer = outputStream.bufferedWriter()
        writer.write(jsonString)
        writer.flush()
    }

    private fun buildCommodityJson(c: Currency): JsonObject = buildJsonObject {
        put(BackupKeys.NAME, c.name)
        put(BackupKeys.POSITION, c.position.toDbString())
        put(BackupKeys.HAS_GAP, c.hasGap)
    }

    private fun buildProfileJson(p: Profile): JsonObject = buildJsonObject {
        put(BackupKeys.NAME, p.name)
        put(BackupKeys.UUID, p.uuid)
        put(BackupKeys.URL, p.url)
        put(BackupKeys.USE_AUTH, p.isAuthEnabled)

        if (p.isAuthEnabled) {
            p.authentication?.user?.let { put(BackupKeys.AUTH_USER, it) }
            p.authentication?.password?.let { put(BackupKeys.AUTH_PASS, it) }
        }

        if (p.apiVersion != API.auto.toInt()) {
            put(BackupKeys.API_VER, p.apiVersion)
        }

        put(BackupKeys.CAN_POST, p.canPost)

        if (p.canPost) {
            val defaultCommodity = p.defaultCommodityOrEmpty
            if (defaultCommodity.isNotEmpty()) {
                put(BackupKeys.DEFAULT_COMMODITY, defaultCommodity)
            }
            put(BackupKeys.SHOW_COMMODITY, p.showCommodityByDefault)
            put(BackupKeys.SHOW_COMMENTS, p.showCommentsByDefault)
            put(BackupKeys.FUTURE_DATES, p.futureDates.toInt())
            put(BackupKeys.PREF_ACCOUNT, p.preferredAccountsFilter)
        }

        put(BackupKeys.COLOUR, p.theme)
    }

    private fun buildTemplateJson(t: TemplateWithAccounts): JsonObject = buildJsonObject {
        put(BackupKeys.UUID, t.header.uuid)
        put(BackupKeys.NAME, t.header.name)
        put(BackupKeys.REGEX, t.header.regularExpression)

        t.header.testText?.let { put(BackupKeys.TEST_TEXT, it) }
        t.header.dateYear?.let { put(BackupKeys.DATE_YEAR, it) }
        t.header.dateYearMatchGroup?.let { put(BackupKeys.DATE_YEAR_GROUP, it) }
        t.header.dateMonth?.let { put(BackupKeys.DATE_MONTH, it) }
        t.header.dateMonthMatchGroup?.let { put(BackupKeys.DATE_MONTH_GROUP, it) }
        t.header.dateDay?.let { put(BackupKeys.DATE_DAY, it) }
        t.header.dateDayMatchGroup?.let { put(BackupKeys.DATE_DAY_GROUP, it) }
        t.header.transactionDescription?.let { put(BackupKeys.TRANSACTION, it) }
        t.header.transactionDescriptionMatchGroup?.let { put(BackupKeys.TRANSACTION_GROUP, it) }
        t.header.transactionComment?.let { put(BackupKeys.COMMENT, it) }
        t.header.transactionCommentMatchGroup?.let { put(BackupKeys.COMMENT_GROUP, it) }

        put(BackupKeys.IS_FALLBACK, t.header.isFallback)

        if (t.accounts.isNotEmpty()) {
            put(
                BackupKeys.ACCOUNTS,
                buildJsonArray {
                    t.accounts.forEach { add(buildTemplateAccountJson(it)) }
                }
            )
        }
    }

    private fun buildTemplateAccountJson(a: TemplateAccount): JsonObject = buildJsonObject {
        a.accountName?.let { put(BackupKeys.NAME, it) }
        a.accountNameMatchGroup?.let { put(BackupKeys.NAME_GROUP, it) }
        a.accountComment?.let { put(BackupKeys.COMMENT, it) }
        a.accountCommentMatchGroup?.let { put(BackupKeys.COMMENT_GROUP, it) }
        a.amount?.let { put(BackupKeys.AMOUNT, it.toDouble()) }
        a.amountMatchGroup?.let { put(BackupKeys.AMOUNT_GROUP, it) }
        a.negateAmount?.let { put(BackupKeys.NEGATE_AMOUNT, it) }
        a.currency?.let { put(BackupKeys.CURRENCY, it) }
        a.currencyMatchGroup?.let { put(BackupKeys.CURRENCY_GROUP, it) }
    }

    private suspend fun getCommodities(): List<Currency> =
        currencyRepository.getAllCurrenciesAsDomain().getOrElse { emptyList() }

    private suspend fun getProfiles(): List<Profile> = profileRepository.getAllProfiles().getOrElse { emptyList() }

    @Suppress("DEPRECATION")
    private suspend fun getTemplates(): List<TemplateWithAccounts> =
        templateRepository.getAllTemplatesWithAccounts().getOrElse { emptyList() }
}
